File: UAV_Controller_v5/minipack.cpp

#include "minipack.h"



MinipackInput::MinipackInput()
{
  error = false;
  started = false;
}


byte MinipackInput::Give(char ch)
{
  if ( ch == MINIPACK_MESSAGE_START )
  {
    // new message, start filling buffer
    packet[0] = ch;
    started = true;
    index = 1;
    error = false; // Important to note we clear the error flag here.
  }
  
  else if (started) // if we already have a packet to fill
  {
    if (index == MINIPACKET_MAX_PACKET_LENGTH)
    {
      // buffer overflow inbound... ABORT!
      started = false;
      index = 0;
    }
    else
    {
      packet[index++] = ch; // Load 'er up!
      
      if (ch == MINIPACK_MESSAGE_END) // check for termination
      {
        packet[index] = 0; // null terminate for good measure.
        started = false;
        
        byte index_copy = index - 2; // because we dont include the brackets as part of the packet
        index = 1;
        return index_copy;
      }
    }
  }
  return 0;
}

byte MinipackInput::Chr2Bits(char ch)
{
  byte bits = ch - MINIPACK_MESSAGE_ZERO;
  error |= (bits >= MINIPACK_MESSAGE_BASE);
  return bits;
}

uint32_t MinipackInput::Unpack(byte char_count)
{
  if (error) { return 0; } // if there is a logged error, then dont waste any time
  
  uint32_t num = 0;
  for ( byte i = 0; i < char_count; i++ )
  {
    // for each char, we read it, then decode it
    byte bits = Chr2Bits(packet[index + i]);
    
    num <<= MINIPACK_CHAR_BITS;
    num += bits;
  }
  index += char_count;
  return num;
}

int32_t MinipackInput::UnpackSigned(byte char_count)
{
  uint32_t halfway = 1 << ((char_count*MINIPACK_CHAR_BITS) - 1);
  return Unpack(char_count) - halfway; 
}

float MinipackInput::UnpackFloat()
{
  uint32_t inum = 0;
  inum = Chr2Bits(packet[index++]);
  inum <<= 6;
  inum |= Chr2Bits(packet[index++]);
  inum <<= 2;

  byte b3 = Chr2Bits(packet[index++]);
  inum |= (b3 >> 4) & 0x3;

  bool neg = (b3 & 0x8);
  byte scalar = b3 & 0x7;

  float num = inum / MINIPACK_FLOAT_SCALAR_TABLE[scalar];
  num = (neg) ? -num : num;
  return num;
}

bool MinipackInput::UnpackError()
{
  return ( (error) || (packet[index] != MINIPACK_MESSAGE_END) );
}





MinipackOutput::MinipackOutput()
{
  index = 1;
  packet[0] = MINIPACK_MESSAGE_START;
  packet[1] = 0;
}


void MinipackOutput::NewPacket()
{
  index = 1;
}

void MinipackOutput::Pack(byte char_count, uint32_t num)
{
  byte end_index = index + char_count - 1;
  for (byte i = 0; i < char_count; i++)
  {
    byte bits = num & ((1 << MINIPACK_CHAR_BITS) - 1);
    packet[end_index - i] = Bits2Chr(bits);
    num >>= MINIPACK_CHAR_BITS;
  }
  index += char_count;
}

void MinipackOutput::PackSigned(byte char_count, int32_t num)
{
  uint32_t halfway = 1 << ((char_count*MINIPACK_CHAR_BITS) - 1);
  Pack(char_count, halfway + num);
}

void MinipackOutput::PackFloat(float num)
{
  bool neg = num < 0;
  num = (neg) ? -num : num;

  byte i = 0;
  while ((i < MINIPACK_FLOAT_MAX_SCALAR) && (num * MINIPACK_FLOAT_SCALAR_TABLE[i+1] < MINIPACK_FLOAT_MAX_MANTISSA))
  {
    i++;
  }
  
  uint32_t inum = num*MINIPACK_FLOAT_SCALAR_TABLE[i];

  packet[index+2] = Bits2Chr( ((inum & 0x3) << 4) | (neg << 3) | i );
  inum >>= 2;
  packet[index+1] = Bits2Chr( inum & 0x3F );
  inum >>= 6;
  packet[index+0] = Bits2Chr( inum & 0x3F );
  index += 3;
}

char* MinipackOutput::EndPacket()
{
  packet[index] = MINIPACK_MESSAGE_END;
  packet[index + 1] = 0;

  return packet;
}


char MinipackOutput::Bits2Chr(byte bits)
{
  return bits + MINIPACK_MESSAGE_ZERO;
}

File: Python/minipack.py
MINIPACKET_MAX_PACKET_LENGTH = 50

MINIPACK_MESSAGE_START = ord('(')
MINIPACK_MESSAGE_END = ord(')')

MINIPACK_MESSAGE_ZERO = ord('0')
MINIPACK_MESSAGE_BASE = 64
MINIPACK_CHAR_BITS = 6



MINIPACK_FLOAT_CHARS = 3;
MINIPACK_FLOAT_SCALAR_BITS = 3;
MINIPACK_FLOAT_MANTISSA_BITS = (MINIPACK_CHAR_BITS * MINIPACK_FLOAT_CHARS) - MINIPACK_FLOAT_SCALAR_BITS - 1;

MINIPACK_FLOAT_MAX_MANTISSA = (1 << MINIPACK_FLOAT_MANTISSA_BITS) - 1;
MINIPACK_FLOAT_MAX_SCALAR = (1 << MINIPACK_FLOAT_SCALAR_BITS) - 1;

MINIPACK_FLOAT_SCALAR_TABLE = [
  0.01,
  1,
  10,
  100,
  1000,
  10000,
  100000,
  1000000,
]




class MinipackInput():
    def __init__(self):
        self.error = False
        self.started = False
        self.index = 0
        self.packet = [0]*MINIPACKET_MAX_PACKET_LENGTH


    def Give(self, ch):

        ch = ord(ch)

        if ch == MINIPACK_MESSAGE_START:
            self.packet[0] = ch
            self.started = True
            self.index = 1
            self.error = False

        elif self.started:

            if self.index == MINIPACKET_MAX_PACKET_LENGTH:
                self.started = False
                self.index = 0;
            else:
                self.packet[self.index] = ch;
                self.index += 1
                if ch == MINIPACK_MESSAGE_END:
                    self.packet[self.index] = 0;
                    
                    index_copy = self.index - 2
                    self.index = 1
                    return index_copy
        return 0

    
    def Char2Bits(self, ch):
        bits = ch - MINIPACK_MESSAGE_ZERO
        if bits >= MINIPACK_MESSAGE_BASE:
            self.error = True
        return bits

    def Unpack(self, char_count):

        if self.error: return 0

        num = 0;
        for i in range(char_count):
            bits = self.Char2Bits( self.packet[self.index + i] )
            num <<= MINIPACK_CHAR_BITS
            num += bits

        self.index += char_count
        return num

    
    def UnpackSigned(self, char_count):
        halfway = 1 << ((char_count*MINIPACK_CHAR_BITS) - 1);

        return self.Unpack(char_count) - halfway; 

    def UnpackFloat(self):
        inum = 0;
        inum = self.Char2Bits(self.packet[self.index]);
        inum <<= 6;
        inum |= self.Char2Bits(self.packet[self.index+1]);
        inum <<= 2;

        b3 = self.Char2Bits(self.packet[self.index+2]);
        inum |= (b3 >> 4) & 0x3;

        
        self.index += 3

        neg = (b3 & 0x8);
        scalar = b3 & 0x7;

        num = float(inum) / MINIPACK_FLOAT_SCALAR_TABLE[scalar];
        if neg: num = -num;

        return num;

    def UnpackError(self):
        return (self.error) or (self.packet[self.index] != MINIPACK_MESSAGE_END)



class MinipackOutput():
    def __init__(self):
        self.index = 1
        self.packet = [0]*MINIPACKET_MAX_PACKET_LENGTH
        self.packet[0] = MINIPACK_MESSAGE_START

    def NewPacket(self):
        self.index = 1

    def Pack(self, char_count, num):
        end_index = self.index + char_count - 1
        for i in range(char_count):
            bits = num & ((1 << MINIPACK_CHAR_BITS)-1)
            self.packet[end_index - i] = self.Bits2Chr(bits)
            num >>= MINIPACK_CHAR_BITS
        self.index += char_count

    def PackSigned(self, char_count, num):
        halfway = 1 << ((char_count*MINIPACK_CHAR_BITS) - 1);
        self.Pack(char_count, halfway + num);


    def PackFloat(self, num):
        neg = num < 0;
        if (neg): num = -num;
        
        i = 0;
        while (i < MINIPACK_FLOAT_MAX_SCALAR) and (num * MINIPACK_FLOAT_SCALAR_TABLE[i+1] < MINIPACK_FLOAT_MAX_MANTISSA):
          i += 1
        
        #print(i)

        inum = int( num*MINIPACK_FLOAT_SCALAR_TABLE[i] );

        self.packet[self.index+2] = self.Bits2Chr( ((inum & 0x3) << 4) | (neg << 3) | i );
        inum >>= 2;
        self.packet[self.index+1] = self.Bits2Chr( inum & 0x3F );
        inum >>= 6;
        self.packet[self.index+0] = self.Bits2Chr( inum & 0x3F );
        self.index += 3;


    def EndPacket(self):
        self.packet[self.index] = MINIPACK_MESSAGE_END
        self.packet[self.index + 1] = 0
        return self.Packet2String(self.packet)

    def Bits2Chr(self, bits):
        return bits + MINIPACK_MESSAGE_ZERO

    def Packet2String(self, packet):
        outgoingpacket = []
        i = 0
        ch = self.packet[i]
        while ch != 0:
            outgoingpacket.append( chr(ch) )
            i += 1
            ch = self.packet[i]

        return "".join(outgoingpacket)



